File: 421.数组中两个数的最大异或值.cpp
#include <vector>
#include <unordered_set>
using namespace std;
/*
 * @lc app=leetcode.cn id=421 lang=cpp
 *
 * [421] 数组中两个数的最大异或值
 */

 // @lc code=start
class Solution {
    public:
    int findMaximumXOR(vector<int>& nums)
    {
        int x = 0;
        for (int k = 30; k >=0; k--)
        {
            unordered_set<int> seen;
            for (int num : nums)
            {
                seen.insert(num >> k);
            }
            int x_next = x * 2 + 1;
            bool found = false;
            for (int num : nums)
            {
                if (seen.count(num>>k^x_next))
                {
                    found = true;
                    break;
                }
                   
            }
            if (found)
            {
                x = x_next;
            }
            else
            {
                x = x_next - 1;
            }
        }
        return x;
    }
};
// @lc code=end

File: 421.数组中两个数的最大异或值.go
package leetcode

/*
 * @lc app=leetcode.cn id=421 lang=golang
 *
 * [421] 数组中两个数的最大异或值
 */

// @lc code=start
func findMaximumXOR(nums []int) (x int) {
	for k := 30; k >= 0; k-- {
		seen := make(map[int]bool)
		for _, num := range nums {
			seen[num>>k] = true
		}
		xNext := x*2 + 1
		found := false

		for _, num := range nums {
			if seen[num>>k^xNext] {
				found = true
				break
			}
		}
		if found {
			x = xNext
		} else {
			x = xNext - 1
		}
	}
	return x
}

// @lc code=end
